proposed/tracing.md

```diff
@@ -10,14 +10,103 @@ interfaces and functionality first.
 
 [RFC 2119]: http://tools.ietf.org/html/rfc2119
 
-## Goal
+## 1. Definitions
 
-TBD
+* **Framework** - An application framework (or micro-framework) that runs a developers code.  
+  Eg. Laravel, Symphony, CakePHP, Slim
+* **Library** - Any library that developers may include that adds additional functionality.  
+  Eg. Image Manipulation, HTTP Clients, 3rd Party SDKs
+* **Provider** - An organization that offers APM (Application Performance Monitoring) as a service.
+  Typically, via a composer package
 
-## Definitions
+## 2. Interfaces
 
-* **Framework** - An application framework (or micro-framework) that runs a developers code.  
-  Eg. Laravel, Symfony, CakePHP, Slim
-* **Library** - Any library that developers may include that adds additional functionality  
-  Eg. Image Manipulation, HTTP Clients, 3rd Party SDK's 
-* **Provider** - An organization that offers APM as a service. Typically via a composer package
+## 2.1 Span
+
+See: [AAllport/psr-tracing - SpanInterface.php](https://github.com/AAllport/psr-tracing/blob/main/src/SpanInterface.php)
+
+A span represents a single operation within a trace. A trace is a collection of spans.
+A fully formed span will consist of the following data:
+
+- Name
+- Attributes, an associative array representing key-value pairs
+- Start Time
+- End Time
+- Outcome
+- Exceptions
+
+Upon creation, a new span will not be activated or have any attributes set.
```

Comment:
I think it's good to keep mutability to a minimum on value objects.
How about?:
- activate the span and start time as constructor time
- have an ending method to fill in the outcome and eventual exceptions, end the time and deactivate the span.
  - I used something in a wrapper like `end()` and `endWithError($exception)` to accomplish this.

Reply:
Activate and start are two different operations. In the majority of cases they are executed at the same time and immediately after the construction, but not always:
For example, you can just prepare an HTTP request via a client and construct the span but without firing it or processing it (no start nor activate should be called).
Then, when sending the request, you should call the `start` method, but you may not call `activate` because the request is in progress in background. Only when you need to wait and process the response you should call `activate` (which marks the span as the currently active span in the context).

The same applies for the outcome and exceptions: adding an exception does not mean that the outcome of the operation is not OK and does not imply the ending of the span; just like ending a span without adding any exception does not imply that the operation has been completed without errors.

Outcomes, exceptions and start/end times are completely unrelated concepts and, unfortunately, simplifying them leaves out some useful cases.